sdk/src/content_credential.rs

```diff
@@ -0,0 +1,328 @@
+    pub fn open_stream(
+        mut self,
+        format: &str,
+        mut stream: impl Read + Seek + Send,
+    ) -> Result<Self> {
+        stream.rewind()?;
+        let ingredient_uri =
+            self.add_ingredient_from_stream(Relationship::ParentOf, format, &mut stream)?;
+        self.add_action(Action::new(c2pa_action::OPENED).add_ingredient(ingredient_uri)?)?;
+        Ok(self)
+    }
+
+    fn parent_ingredient(&self) -> Option<Ingredient> {
+        for i in self.claim.ingredient_assertions() {
+            if let Ok(ingredient) = Ingredient::from_assertion(i.assertion()) {
+                if ingredient.relationship == Relationship::ParentOf {
+                    return Some(ingredient);
+                }
+            }
+        }
+        None
+    }
+
+    /// adds an assertion to the content credential's claim
+    pub fn add_assertion(&mut self, assertion: &impl AssertionBase) -> Result<HashedUri> {
+        self.claim.add_assertion(assertion)
+    }
+
+    /// adds an ingredient assertion to the content credential's claim
+    /// if manifest_bytes is provided, it will be loaded into the store
+    pub fn add_ingredient_assertion(
+        &mut self,
+        ingredient_assertion: &Ingredient,
+        manifest_bytes: Option<&[u8]>,
+    ) -> Result<HashedUri> {
+        if let Some(manifest_bytes) = manifest_bytes {
+            let store = Store::load_ingredient_to_claim(
+                &mut self.claim,
+                manifest_bytes,
+                None,
+                self.context,
+            )?;
+            if ingredient_assertion.relationship == Relationship::ParentOf {
+                // we must replace the store for parent ingredients
+                self.store = store;
+            }
+        }
+        // add the ingredient assertion and get it's uri
+        self.add_assertion(ingredient_assertion)
+    }
+
+    /// create a manifest store report from the store and validation results
+    pub fn add_action(&mut self, action: Action) -> Result<()> {
+        self.claim.add_action(action)?;
+        Ok(())
+    }
+
+    /// internal implementation to add ingredient from stream
+    fn add_ingredient_from_stream(
+        &mut self,
+        relationship: Relationship,
+        format: &str,
+        mut stream: impl Read + Seek + Send,
+    ) -> Result<HashedUri> {
+        let (ingredient_assertion, manifest_bytes) =
+            Ingredient::from_stream(relationship.clone(), format, &mut stream, self.context)?;
+
+        self.add_ingredient_assertion(&ingredient_assertion, manifest_bytes.as_deref())
+    }
+
+    /// signs and saves the content credential to the destination stream
+    pub fn save_to_stream<R, W>(
+        &mut self,
+        format: &str,
+        source: &mut R,
+        dest: &mut W,
+    ) -> Result<Vec<u8>>
+    where
+        R: Read + Seek + Send,
+        W: Write + Read + Seek + Send,
+    {
+        let signer = self.context.signer()?;
```

Comment:
I think that's the ebst approach. Let's a user also define a signer and put it in the context,